### PR TITLE
5. kubectl-kcp: get rid of workspace duality and implement logical cluster path semantics

### PR DESCRIPTION
**Pull request by sttts**
Implementing UX from https://docs.google.com/document/d/1XJJCr16bg22QuJnQB2W98A-djRTtrLr2JE2IWZUHzJU/edit#heading=h.nn8danthd1ys.

- simplified and complete kubectl plugin to implement the second section in https://docs.google.com/document/d/1XJJCr16bg22QuJnQB2W98A-djRTtrLr2JE2IWZUHzJU/edit#heading=h.nn8danthd1ys.
- removed duality between v1beta1 `Workspaces` in some virtual workspace path and the `ClusterWorkspaces` in an org. Now you can do `kubectl get workspaces` in an org

Based on https://github.com/kcp-dev/kcp/pull/744.

### CODE REVIEW

File: pkg/cliplugins/workspace/plugin/kubeconfig_test.go [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+				Clusters:  map[string]*clientcmdapi.Cluster{"test": {Server: "https://test/clusters/root:foo"}},
+				AuthInfos: map[string]*clientcmdapi.AuthInfo{"test": {Token: "test"}},
+			},
+			existingWorkspaces: []string{"bar"},
+			newWorkspaceName:   "bar",
+			useAfterCreation:   true,
+			wantErr:            true,
+		},
+		{
+			name: "create, use after creation, ignore existing",
```
Comment (ncdc) [inline pkg/cliplugins/workspace/plugin/kubeconfig_test.go:114]: I don't think this test case is testing what the description says? It's missing `existingWorkspaces: []string{"bar"}`.

File: pkg/cliplugins/workspace/plugin/kubeconfig.go [hunk 2]
```diff
@@ -20,60 +20,83 @@ import (
...
-	"github.com/kcp-dev/kcp/pkg/virtual/workspaces/registry"
 )
 
 const (
-	kcpPreviousWorkspaceContextKey         string = "workspace.kcp.dev/previous"
-	kcpCurrentWorkspaceContextKey          string = "workspace.kcp.dev/current"
-	kcpOverridesContextAuthInfoKey         string = "workspace.kcp.dev/overridden-auth"
-	kcpVirtualWorkspaceInternalContextName string = "workspace.kcp.dev/workspace-directory"
+	kcpPreviousWorkspaceContextKey string = "workspace.kcp.dev/previous"
+	kcpCurrentWorkspaceContextKey  string = "workspace"
```
Comment (ncdc) [inline pkg/cliplugins/workspace/plugin/kubeconfig.go:43]: Curious: why did you rename from `workspace.kcp.dev/current` to `workspace`?

File: pkg/cliplugins/workspace/cmd/cmd.go [hunk 3]
```diff
@@ -29,126 +29,131 @@ import (
...
-			if err := kubeconfig.CreateWorkspace(c.Context(), opts, args[0], useAfterCreation, typeFlag); err != nil {
+			if err := kubeconfig.CreateWorkspace(cmd.Context(), args[0], workspaceType, ignoreExisting, enterAfterCreation); err != nil {
 				return err
 			}
 			return nil
 		},
 	}
-	createCmd.Flags().Bool(useFlag, false, "Use the new workspace after a successful creation")
-	createCmd.Flags().StringVar(&typeFlag, "type", typeFlag, "The workspace type to create")
+	createCmd.Flags().StringVar(&workspaceType, "type", "", "A workspace type (default: Universal)")
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/cliplugins/workspace/cmd/cmd.go:141]: We could. This is set by defaulting in the API. I didn't want to stop us from having a default by workspace, e.g. in the root it would be Org, in the org it would be Universal.

[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/cliplugins/workspace/cmd/cmd.go:141]: I see what you're saying. I hesitate any time I see manual "default: xyz" in a flag description, as it usually comes from whatever default value you specify in the flag definition.

File: pkg/cliplugins/workspace/plugin/helpers.go [hunk 4]
```diff
@@ -29,11 +29,34 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	virtualcommandoptions "github.com/kcp-dev/kcp/cmd/virtual-workspaces/options"
-	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
 	tenancyv1beta1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1beta1"
 	tenancyclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
 )
 
+func parseClusterURL(host string) (*url.URL, logicalcluster.LogicalCluster, error) {
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/cliplugins/workspace/plugin/helpers.go:37]: added

File: pkg/server/handler.go [hunk 5]
```diff
@@ -145,6 +148,27 @@ func WithClusterScope(apiHandler http.Handler) http.HandlerFunc {
...
+		cluster := genericapirequest.ClusterFrom(req.Context())
+		if cluster.Name.Empty() {
+			apiHandler.ServeHTTP(w, req)
+			return
+		}
+
+		toRedirectPath := path.Join("/apis", tenancyv1beta1.SchemeGroupVersion.Group, tenancyv1beta1.SchemeGroupVersion.Version, "workspaces/")
+		if strings.HasPrefix(req.URL.Path, toRedirectPath) {
+			newPath := path.Join("/services/workspaces", cluster.Name.String(), "personal", req.URL.Path)
+			klog.Infof("Rewriting %s -> %s", path.Join(cluster.Name.Path(), req.URL.Path), newPath)
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/server/handler.go:164]: 4. Done

File: pkg/cliplugins/workspace/plugin/kubeconfig_test.go (hunk 1 above)
Comment (sttts) [reply pkg/cliplugins/workspace/plugin/kubeconfig_test.go:114]: done

File: pkg/cliplugins/workspace/plugin/kubeconfig.go (hunk 2 above)
Comment (sttts) [reply pkg/cliplugins/workspace/plugin/kubeconfig.go:43]: `workspace.kcp.dev/current` appeared to be very cryptic, and one might use the workspace context manually. No strong opinion.

Related: https://github.com/kcp-dev/kcp/pull/779

Comment (sttts) [reply pkg/cliplugins/workspace/plugin/kubeconfig.go:43]: @davidfestal @ncdc opinions?

File: pkg/cliplugins/workspace/plugin/kubeconfig.go [hunk 6]
```diff
@@ -82,468 +105,215 @@ func NewKubeConfig(opts *Options) (*KubeConfig, error) {
...
-	kc.startingConfig.Contexts[kcpCurrentWorkspaceContextKey] = &api.Context{
-		Cluster:  kcpCurrentWorkspaceContextKey,
-		AuthInfo: authInfoName,
+	// modify kubeconfig, using the "workspace" context and cluster
+	newKubeConfig := kc.startingConfig.DeepCopy()
+	oldCluster, found := kc.startingConfig.Clusters[currentContext.Cluster]
+	if !found {
+		return fmt.Errorf("cluster %q not found in kubeconfig", currentContext.Cluster)
 	}
+	newCluster := *oldCluster
```
Comment (davidfestal) [inline pkg/cliplugins/workspace/plugin/kubeconfig.go:210]: So you don't manage AuthInfo at all in the current and previous context ?
Afair in previous tests I had done, it might bring some very painful behavior where you cannot bring back the auth used (through a --token override for example) with the previous workspace.

File: pkg/cliplugins/workspace/plugin/kubeconfig.go [hunk 7]
```diff
@@ -82,468 +105,215 @@ func NewKubeConfig(opts *Options) (*KubeConfig, error) {
...
-		return result.Error()
-	}
-
-	if statusCode != http.StatusOK {
-		rawResult, err := result.Raw()
-		if err != nil {
-			return err
+	parentClusterName, workspaceName := clusterName.Split()
+	workspacePrettyName := workspaceName
+	if !parentClusterName.Empty() && clusterName.HasPrefix(tenancyv1alpha1.RootCluster) {
```
Comment (davidfestal) [inline pkg/cliplugins/workspace/plugin/kubeconfig.go:253]: Not sure I understand the meaning of this condition:
``` golang
&& clusterName.HasPrefix(tenancyv1alpha1.RootCluster)
```

File: pkg/cliplugins/workspace/plugin/kubeconfig.go (hunk 2 above)
Comment (ncdc) [reply pkg/cliplugins/workspace/plugin/kubeconfig.go:43]: I think I prefer the longer form, but I don't know we need to do too much bikeshedding since #779 should make this more user-driven?

File: pkg/cliplugins/workspace/plugin/kubeconfig.go (hunk 7 above)
Comment (ncdc) [reply pkg/cliplugins/workspace/plugin/kubeconfig.go:253]: To only deal with root:.... and not system:...?

File: pkg/cliplugins/workspace/plugin/kubeconfig.go (hunk 6 above)
Comment (ncdc) [reply pkg/cliplugins/workspace/plugin/kubeconfig.go:210]: Yes I think this needs to be carried forward

File: pkg/cliplugins/workspace/plugin/kubeconfig.go (hunk 7 above)
Comment (sttts) [reply pkg/cliplugins/workspace/plugin/kubeconfig.go:253]:
```
	parentClusterName, workspaceName := clusterName.Split()
	workspacePrettyName := workspaceName
	if clusterName.HasPrefix(tenancyv1alpha1.RootCluster) && clusterName != tenancyv1alpha1.RootCluster {
		ws, err := getWorkspaceFromInternalName(ctx, workspaceName, kc.client.Cluster(parentClusterName))
		if err == nil {
			workspacePrettyName = ws.Name
		}
	}
```

File: pkg/cliplugins/workspace/plugin/kubeconfig.go (hunk 2 above)
Comment (sttts) [reply pkg/cliplugins/workspace/plugin/kubeconfig.go:43]: switched back to the longer form.

File: pkg/cliplugins/workspace/cmd/cmd.go [hunk 8]
```diff
@@ -29,65 +30,85 @@ import (
...
+		if len(args) > 1 {
+			return cmd.Help()
+		}
+
+		arg := ""
+		if len(args) == 1 {
+			arg = args[0]
+		} else if len(args) > 1 {
+			return cmd.Help()
+		}
```
Comment (ncdc) [inline pkg/cliplugins/workspace/cmd/cmd.go:77]: Dupe of L68-70

File: pkg/cliplugins/workspace/cmd/cmd.go [hunk 9]
```diff
@@ -29,65 +30,85 @@ import (
...
+		return kubeconfig.UseWorkspace(cmd.Context(), arg)
+	}
+
 	cmd := &cobra.Command{
 		Aliases:          []string{"ws", "workspaces"},
-		Use:              "workspace [--scope=<personal|all>] <create|list|use|current|delete>",
+		Use:              "workspace [create|<workspace>|..|-|<root:absolute:workspace>]",
```
Comment (ncdc) [inline pkg/cliplugins/workspace/cmd/cmd.go:83]: Need to add `list` back here?

File: pkg/cliplugins/workspace/plugin/kubeconfig.go (hunk 6 above)
Comment (ncdc) [reply pkg/cliplugins/workspace/plugin/kubeconfig.go:219]: Based on @sttts 's comment in Slack
>verified that overrides don't affect  what is persisted. This is what kubectl itself does too. overrides are only used to tweak a client used by kubectl/the plugin.

👍

File: pkg/cliplugins/workspace/plugin/kubeconfig.go [hunk 10]
```diff
@@ -188,7 +193,7 @@ func (kc *KubeConfig) UseWorkspace(ctx context.Context, name string) error {
 			newServerHost = u.String()
 		} else {
 			// relative logical cluster, get URL from workspace object in current context
-			ws, err := kc.client.Cluster(currentClusterName).TenancyV1beta1().Workspaces().Get(ctx, name, metav1.GetOptions{})
+			ws, err := kc.clusterClient.Cluster(currentClusterName).TenancyV1beta1().Workspaces().Get(ctx, name, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/cliplugins/workspace/plugin/kubeconfig.go:200]: yes, good spot.

File: pkg/cliplugins/workspace/cmd/cmd.go (hunk 8 above)
Comment (sttts) [reply pkg/cliplugins/workspace/cmd/cmd.go:77]: indeed

File: pkg/cliplugins/workspace/cmd/cmd.go (hunk 9 above)
Comment (sttts) [reply pkg/cliplugins/workspace/cmd/cmd.go:83]: done

**Review by ncdc — APPROVED** [review]
